pandas/core/groupby/generic.py

```diff
@@ -470,14 +487,34 @@ def _get_index() -> Index:
             )
             return self._reindex_output(result)
 
-    def _aggregate_named(self, func, *args, **kwargs):
+    def _aggregate_maybe_named(self, func, *args, **kwargs):
+        """
+        Try the named-aggregator first, then unnamed, which better matches
+        what libreduction does.
+        """
+        try:
+            return self._aggregate_named(func, *args, named=True, **kwargs)
```

Comment:
what is this actually doing?

Reply:
trying to track down an answer to why i did this, may take a bit

Reply:
OK now i recall.  in e.g. test_apply_columns_multilevel we do:

```
        cols = pd.MultiIndex.from_tuples([("A", "a", "", "one"), ("B", "b", "i", "two")])
        ind = date_range(start="2017-01-01", freq="15Min", periods=8)
        df = DataFrame(np.array([0] * 16).reshape(8, 2), index=ind, columns=cols)
        agg_dict = {col: (np.sum if col[3] == "one" else np.mean) for col in df.columns}
        result = df.resample("H").apply(lambda x: agg_dict[x.name](x))
```

so the function we are passing to apply is `lambda x: agg_dict[x.name](x)` depends on the Series name `x.name` but what we're actually doing ATM (and since this is tested, i guess it means we support it on purpose?) is patching the Series name to match the group name so that we end up applying a different function to each group.

But the user could also pass a function that depends on the non-patched Series name, so we end up guessing which regime we're in, which this is doing.

If it were up to me I'd shoot this name-patching thing into the sun.

Reply:
grr, this is *really* complex.